canny/canny.py

```diff
@@ -0,0 +1,204 @@
+
+Note on optimising TensorFlow GPU performance:
+ 1. https://www.tensorflow.org/guide/gpu_performance_analysis
+ 2. https://github.com/NVIDIA/DeepLearningExamples/issues/57
+"""
+import tensorflow as tf
+import tensorflow_probability as tfp
+import numpy as np
+import os
+import sys
+
+sys.path.append("../components/")
+import connected_components
+
+os.environ['TF_GPU_THREAD_MODE'] = 'gpu_private'
+
+
+class CannyEdge:
+    """Canny edge detector.
+
+    Extracts edges from grayscale images employing a set of algorithms. The Process of Canny edge detection algorithm
+    can be broken down to 5 different steps:
+     1. Apply Gaussian filter to smooth the image in order to remove the noise.
+     2. Find the intensity gradients of the image.
+     3. Apply gradient magnitude thresholding or lower bound cut-off suppression to get rid of spurious response to
+     edge detection.
+     4. Apply double threshold to determine potential edges.
+     5. Track edge by hysteresis: Finalize the detection of edges by suppressing all the other edges that are weak and
+     not connected to strong edges.
+    """
+
+    def __init__(self, weak_threshold=None, strong_threshold=None):
+        """Initialise class attributes.
+
+        Args:
+            weak_threshold: first threshold for the hysteresis procedure.
+            strong_threshold: second threshold for the hysteresis procedure.
+        """
+        self.weak_threshold = tf.constant(weak_threshold, dtype=tf.float32) if weak_threshold is not None else None
+        self.strong_threshold = tf.constant(strong_threshold, dtype=tf.float32) if weak_threshold is not None else None
+        self.pi = tf.constant(np.pi, dtype=tf.float32)
+
+        # filters for canny edge detection
+        filter_0, filter_90 = np.zeros(shape=(3, 3, 1, 2)), np.zeros(shape=(3, 3, 1, 2))
+        filter_45, filter_135 = np.zeros(shape=(3, 3, 1, 2)), np.zeros(shape=(3, 3, 1, 2))
+
+        filter_0[1, 0, 0, 0], filter_0[1, 2, 0, 1] = 1, 1
+        self.filter_0 = tf.constant(filter_0, tf.float32)
+
+        filter_90[0, 1, 0, 0], filter_90[2, 1, 0, 1] = 1, 1
+        self.filter_90 = tf.constant(filter_90, tf.float32)
+
+        filter_45[0, 0, 0, 0], filter_45[2, 2, 0, 1] = 1, 1
+        self.filter_45 = tf.constant(filter_45, tf.float32)
+
+        filter_135[0, 2, 0, 0], filter_135[2, 0, 0, 1] = 1, 1
+        self.filter_135 = tf.constant(filter_135, tf.float32)
+
+        self.canny_strong = tf.constant(1.33, dtype=tf.float32)
+        self.canny_weak = tf.constant(0.67, dtype=tf.float32)
+
+        # connected components
+        self.connected_components = connected_components.ConnectedComponents()
+
+    @tf.function(input_signature=[tf.TensorSpec(shape=[None, None, None], dtype=tf.float32)])
+    def __call__(self, images):
+        """Finds edges on the images employing Canny edge detection algorithm.
+
+        Args:
+            images: 3D tensor NHW (float32). Should be grayscale images.
```

Comment:
@arvindrajan92 by NHW do you mean Nth Image, then Height, then Width?

Reply:
Yes, just the term on TensorFlow docs. C in NHWC would be Channel.